### ISSUE TITLE
yolov3-tiny: Average loss doesn't reduce during training. Detection also fails.

### ISSUE THREAD

**Issue by hosaka** [AlexeyAB/darknet#1672]
I'm 100% sure i'm doing something wrong here. What I have:
- @AlexeyAB fork, not pjreddie/darknet
- Open Images Dataset v4, 80/20 split for train/validation with labels
- 1 class (person), 5k train images (+1.3k negative images, empty labels), 1k valid images.
- Calculated anchors replaced in cfg, as well as number of classes and filters as per README.md.
- Training on two GPUs.
- Training with yolov3-tiny.conv.15
- Similar setup **works with regular yolov3**, my average loss went to 0.06 after 15k iterations. So the very same labels and images worked there, nothing changed in the dataset itself.

The issue:
During training the current average loss starts fluctuating around certain point. I've read some other issue and saw hints towards reducing learning rate or simply using defaults from yolov3-tiny_obj.cfg. After larger number of iterations I get another spike up and the same pattern repeats, don't have a chart unfortunately. But here's a common one:
![chart](https://user-images.githubusercontent.com/171876/46128380-3703ff00-c26e-11e8-92ec-52c4579245dc.jpg)

What I've tried as per README.md:
- Setting learning_rate to 0.001 / number of GPUs - 0.005 in my case
- Setting burn_in to default * number of GPUs - 2000 in my case
- Training with defaults on 1 GPU first for 1k-2k iterations, loading the weights and continuing with 2 GPUs.

Results:
Running detector map (on the 12k checkpoint weights from the chart above) mAP/IOU = 0%. Of course, detector test doen't work either.
```
detections_count = 0, unique_truth_count = 1762
class_id = 0, name = person,     ap = 0.00 %
 for thresh = 0.25, precision = -nan(ind), recall = 0.00, F1-score = -nan(ind)
 for thresh = 0.25, TP = 0, FP = 0, FN = 1762, average IoU = 0.00 %
```

I'd be thankful is someone could point out my mistake!

Config:
```
[net]
# Testing
# batch=1
# subdivisions=1
# Training
batch=64
subdivisions=4
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 20000
policy=steps
steps=400000,450000
scales=.1,.1

[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

###########

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=18
activation=linear

[yolo]
mask = 3,4,5
# anchors = 10,14,  23,27,  37,58,  81,82,  135,169,  344,319
anchors = 21.0939,46.3084, 52.0696,128.2728, 76.9742,241.3421, 196.6118,211.1451, 154.3483,351.4461, 325.7273,366.3307
classes=1
num=6
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 8

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=18
activation=linear

[yolo]
mask = 0,1,2
# anchors = 10,14,  23,27,  37,58,  81,82,  135,169,  344,319
anchors = 21.0939,46.3084, 52.0696,128.2728, 76.9742,241.3421, 196.6118,211.1451, 154.3483,351.4461, 325.7273,366.3307
classes=1
num=6
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1
```

**Comment by AlexeyAB** [comment]
* What GPU, CUDA and cuDNN versions do you use?

* What params do you use in the Makefile?

* use CUDNN_HALF=0

* Try to use for training - if you use CUDNN=1 then try CUDNN=0, or vice versa.

**Comment by hosaka** [comment]
@AlexeyAB Hi Alexey, thank you for taking time to answer.
- GTX 1070 8Gb and GTX 980 Ti, CUDA 9.1, cuDNN 7.1.3 (for CUDA 9.1), opencv 3.4.0
- I haven't modified the Makefile, and didn't pass CUDNN_HALF flag either.

I will try to recompile with both of your suggestions. Большое спасибо!

**Comment by AlexeyAB** [comment]
So try to build without CUDNN then: open `\darknet.sln` -> (right click on project) -> properties -> C/C++ -> Preprocessor -> Preprocessor Definitions, and remove there: `CUDNN;`

And try to train by using GTX 980 Ti - use flag `-gpus 1` at the end of training command, if GTX 980 Ti is under number 1

**Comment by hosaka** [comment]
Yep, just tried recompiling without the CUDNN flag:
Training with both GPUs or only 980 Ti, avg loss starts high and continues increasing until it quickly hits `nan`

With CUDNN:
Using only 980 Ti, doesn't look like the performance changes at all, it still fluctuates between 2.0~2.7:
![chart](https://user-images.githubusercontent.com/171876/46161173-c685ce00-c2bf-11e8-8c2f-22e17663e7f2.jpg)

I will check once again that my dataset is in order. But I'm surprised by why it trained fine the first time on the regular yolov3.

**Comment by AlexeyAB** [comment]
* Check files `bad.list` and `bad_label.list`
* Check your dataset by using https://github.com/AlexeyAB/Yolo_mark

**Comment by hosaka** [comment]
I re-wrote Open Images Dataset fetching code, double checked the label calculation and built a new dataset. The issue has disappeared. It must have been my own error, even though the labels I've had, were matching the labels I'd generate with Yolo_mark. Something must have gone wrong the first time. Thank you for your help, it was my own mistake in the end.
